angular/projects/spark-angular/src/lib/components/sprk-autocomplete/sprk-autocomplete.component.ts

```diff
@@ -0,0 +1,425 @@
+
+      this.closedEvent.emit();
+      this.isOpen = false;
+    }
+  }
+
+  /**
+   * Show the Autocomplete results list.
+   */
+  showResults(): void {
+    if (this.results) {
+      this.renderer.removeClass(
+        this.results.nativeElement,
+        'sprk-c-Autocomplete__results--hidden',
+      );
+
+      if (this.input) {
+        // Set aria-expanded on the input
+        this.renderer.setAttribute(
+          this.input.ref.nativeElement.parentNode,
+          'aria-expanded',
+          'true',
+        );
+      }
+
+      this.openedEvent.emit();
+      this.isOpen = true;
+    }
+  }
+
+  /**
+   * @ignore
+   * Track the index of the currently-highlighted Autocomplete result item.
+   */
+  highlightedIndex = -1;
+
+  /**
+   * @ignore
+   */
+  ngAfterContentInit(): void {
+    if (this.results) {
+      if (this.isOpen) {
+        this.showResults();
+      } else {
+        this.hideResults();
+      }
+
+      if (this.input) {
+        this.generateAriaControls(
+          this.input.ref.nativeElement,
+          this.results.nativeElement,
+        );
+        this.generateAriaOwns(
+          this.input.ref.nativeElement.parentNode,
+          this.results.nativeElement,
+        );
+      }
+    }
+
+    // if itemSelectedEvent is specified, also set that as the click event on each result
+    if (this.itemSelectedEvent) {
+      this.resultItems.forEach((element, index) => {
+        element.clickedEvent = this.itemSelectedEvent;
+      });
+    }
+  }
+
+  /**
+   * @ignore
+   * This logic needs to happen in this event so that the input width is
+   * correctly calculated.
+   */
+  ngAfterViewInit(): void {
+    if (this.input) {
+      this.calculateResultsWidth();
+    }
+  }
+
+  /**
+   * @ignore
+   * Set the max-width on the Autocomplete results equal to the current
+   * width of the input element.
+   */
+  calculateResultsWidth() {
+    if (this.input) {
+      const currentInputWidth = this.input.ref.nativeElement.offsetWidth;
+
+      if (this.results) {
+        this.renderer.setAttribute(
+          this.results.nativeElement,
+          'style',
+          'max-width:' + currentInputWidth + 'px',
+        );
+      }
+    }
+  }
+  /** @ignore */
+  isUpPressed = (e) => e.key === 'ArrowUp' || e.keyCode === 38;
+  /** @ignore */
+  isDownPressed = (e) => e.key === 'ArrowDown' || e.keyCode === 40;
+  /** @ignore */
+  isEnterPressed = (e) => e.key === 'Enter' || e.keyCode === 13;
+  /** @ignore */
+  isEscapePressed = (e) =>
+    e.key === 'Escape' || e.key === 'Esc' || e.keyCode === 27;
+
+  /**
+   * @ignore
+   * Copy the value of the id attribute on contentElement
+   * into the aria-controls attribute on triggerElement.
+   * Generate a unique ID if needed.
+   * @param triggerElement The element that will receive the aria-controls attribute.
+   * @param contentElement The element that will receive an id if needed.
+   */
+  generateAriaControls = (triggerElement, contentElement) => {
```

Comment:
`generateAriaControls` and `generateAriaOwns` are utility functions in HTML, but we've never needed them in Angular before. These can be refactored into utility functions if they're needed anywhere else.